metpy/calc/tools.py

```diff
@@ -330,3 +332,91 @@ def log_interp(x, xp, fp, **kwargs):
+    specified in terms of pressure or height above the bottom of the layer. If the top and
+    bottom of the layer are not in the data, they are interpolated by default.
+
+    Parameters
+    ----------
+    p : array-like
+        Atmospheric pressure profile
+    datavar : array-like
+        Atmospheric variable measured at the given pressures
+    bottom : `pint.Quantity`
+        The bottom of the layer as a pressure or height above the surface pressure
+    depth : `pint.Quantity`
+        The thickness of the layer as a pressure or height above the bottom of the layer
+    interpolate : bool
+        Interpolate the top and bottom points if they are not in the given data
+
+    Returns
+    -------
+    `pint.Quantity, pint.Quantity`
+        The pressure and data variable of the layer
+
+    """
+    # Make sure pressure and datavar are the same length
+    if len(p) != len(datavar):
+        raise ValueError('Pressure and data variable must have the same length.')
+
+    # If no bottom is specified, use the first pressure value
+    if not bottom:
+        bottom_pressure = p[0]
+
+    # If bottom is specified as a height (AGL), convert to pressure
+    if bottom:
+        if bottom.dimensionality == {'[length]': 1.0}:
+            bottom_pressure = add_height_to_pressure(p[0], bottom)
+        if bottom.dimensionality == {'[length]': -1.0, '[mass]': 1.0, '[time]': -2.0}:
+            bottom_pressure = bottom
+
+    # Calculate the pressure at the top of the layer
+    if depth.dimensionality == {'[length]': 1.0}:
+        top_pressure = add_height_to_pressure(bottom_pressure, depth)
+    else:
+        top_pressure = bottom_pressure - depth
+
+    # Handle top or bottom values that are invalid
+    if bottom_pressure > p[0]:
+        raise ValueError(
+            'Bottom of layer pressure is greater than the largest pressure in data.')
+    if top_pressure < p[-1]:
+        raise ValueError('Top of layer pressure is less than the lowest pressure in data.')
+    if bottom_pressure < top_pressure:
+        raise ValueError(
+            'Pressure at the top of the layer is greater than that at the bottom.')
+
+    # Ensure that things are sorted in ascending order
+    sort_inds = np.argsort(p)
+    p = p[sort_inds]
+    datavar = datavar[sort_inds]
+
+    # Mask the pressure values we have data for in the layer
+    inds = (p <= bottom_pressure) & (p >= top_pressure)
+    p_interp = p[inds]
+
+    if interpolate:
+        # If we don't have the bottom or top requested, append them
+        if top_pressure not in p_interp:
+            p_interp = np.sort(np.append(p_interp, top_pressure)) * units.hPa
+        if bottom_pressure not in p_interp:
+            p_interp = np.sort(np.append(p_interp, bottom_pressure)) * units.hPa
+
+        # Interpolate for the possibly missing bottom/top pressure values
+        datavar_interp = log_interp(p_interp, p, datavar).to(units.degC)
```

Comment:
Forcing a return in degC. Should be return whatever units are passed in for temperature and pressure, or stick with hPa, degC?